rs/nervous_system/tools/release-runscript/src/commit_switcher.rs

```diff
@@ -0,0 +1,118 @@
+            .arg("rev-parse")
+            .arg("HEAD")
+            .output()?;
+        let from_commit = String::from_utf8(from_commit.stdout)?.trim().to_string();
+
+        // Check for local changes
+        let status = Command::new("git")
+            .current_dir(&ic)
+            .args(["status", "--porcelain"])
+            .output()?;
+        let has_changes = !status.stdout.is_empty();
+
+        // stash if we have changes
+        if has_changes {
+            println!("{}", format!("Stashing changes...").bright_blue());
+            let stash = Command::new("git").current_dir(&ic).arg("stash").output()?;
+            if !stash.status.success() {
+                return Err(
+                    anyhow::anyhow!("{}", String::from_utf8_lossy(&stash.stderr))
+                        .context("Failed to stash changes."),
+                );
+            }
+
+            std::thread::sleep(std::time::Duration::from_secs(1));
+        }
+
+        // switch to the new commit
+        println!(
+            "{}",
+            format!("Switching to commit: {}", to_commit).bright_blue()
+        );
+        let output = Command::new("git")
+            .current_dir(&ic)
+            .arg("switch")
+            .arg("-d")
+            .arg(&to_commit)
+            .output()?;
+        if !output.status.success() {
+            return Err(
+                anyhow::anyhow!("{}", String::from_utf8_lossy(&output.stderr))
+                    .context("Failed to checkout commit."),
+            );
+        }
+
+        Ok(Self {
+            from_commit,
+            stashed: has_changes,
+        })
+    }
+}
+
+impl Drop for CommitSwitcher {
+    fn drop(&mut self) {
+        let ic = ic_dir();
+
+        // reset ic/Cargo.lock, in case people's rust-analyzers have messed it up
+        let _ = Command::new("git")
+            .current_dir(&ic)
+            .arg("reset")
+            .arg("Cargo.lock")
+            .output();
+
+        // switch
+        println!(
+            "{}",
+            format!("Switching back to commit: {}", self.from_commit).bright_blue()
+        );
+        let switch = Command::new("git")
+            .current_dir(&ic)
+            .arg("switch")
+            .arg("-d")
+            .arg(&self.from_commit)
+            .output()
+            .unwrap();
+
+        std::thread::sleep(std::time::Duration::from_secs(1));
+
+        if !switch.status.success() {
+            println!(
+                "{}",
+                format!(
+                    "Failed to switch back to commit. Try running `git switch -d {}`",
+                    self.from_commit
+                )
+                .bright_red()
+            );
+            println!("error: {}", String::from_utf8_lossy(&switch.stderr));
+            return;
+        }
+
+        if self.stashed {
+            // apply stash
+            Command::new("git")
+                .current_dir(&ic)
+                .arg("stash")
+                .arg("apply")
```

Comment:
you'll actually want 'pop' instead of 'apply' as 'apply' doesn't roll back the stash state to what it was previously, which means this operation would then interfere with whatever else the user is doing with git at the time.

Reply:
Done!